Add grain reference documentation and improve grain identity documentation (#38700)
dotnet/docs: docs/orleans/grains/grain-identity.md

---
title: Grain identity
description: Learn about grain identities in .NET Orleans.
ms.date: 12/12/2023
---

# Grain identity

Grains in Orleans each have a single, unique, user-defined identifier which consists of two parts:

1. The grain _type_ name, which uniquely identifies the grain class.
2. The grain _key_, which uniquely identifies a logical instance of that grain class.

The grain type and key are both represented as human-readable strings in Orleans and, by convention, the grain identity is written with the grain type and key separated by a `/` character. For example, `shoppingcart/bob65` represents the grain type named `shoppingcart` with a key `bob65`.

> It is not common to construct grain identities directly. Instead, it is more common to create [grain references](./grain-references.md) using <xref:Orleans.IGrainFactory?displayProperty=nameWithType>.

Reviewer: ```suggestion
It's not common to construct grain identities directly. Instead, it's more common to create [grain references](./grain-references.md) using <xref:Orleans.IGrainFactory?displayProperty=nameWithType>.
```

Author: I need to get used to using contractions liberally

The following sections discuss grain type names and grain keys in more detail.

## Grain type names

> Orleans will create a grain type name for you based on your grain implementation class by removing the suffix "Grain" from the class name, if it's present, and converting the resulting string into its lower-case representation. For example, a class named `ShoppingCartGrain` will be given the grain type name `shoppingcart`. It is recommended that grain type names and keys consist only of printable characters such as alpha-numeric (`a`-`z`, `A`-`Z`, and `0`-`9`) characters and symbols such as `-`, `_`, `@`, `=`. Other characters may or may not be supported and will often need special treatment when printed in logs or appearing as identifiers in other systems such as databases.

Reviewer: ```suggestion
Orleans creates a grain type name for you based on your grain implementation class by removing the suffix "Grain" from the class name, if it's present, and converting the resulting string into its lower-case representation. For example, a class named `ShoppingCartGrain` will be given the grain type name `shoppingcart`. It's recommended that grain type names and keys consist only of printable characters such as alpha-numeric (`a`-`z`, `A`-`Z`, and `0`-`9`) characters and symbols such as `-`, `_`, `@`, `=`. Other characters may or may not be supported and will often need special treatment when printed in logs or appearing as identifiers in other systems such as databases.
```

Alternatively, the <xref:Orleans.Runtime.GrainTypeAttribute?displayProperty=nameWithType> attribute can be used to customize the grain type name for the grain class which it is attached to, as in the following example:

```csharp
[GrainType("cart")]
public class ShoppingCartGrain : IShoppingCartGrain
{
    // Add your grain implementation here
}
```

> In the above example, the grain class, `ShoppingCartGrain` has a grain type name of `cart`. Each grain can only have one grain type name.

Reviewer: ```suggestion
In the preceding example, the grain class, `ShoppingCartGrain` has a grain type name of `cart`. Each grain can only have one grain type name.
```
